upf/unittest.c

```diff
@@ -0,0 +1,299 @@
+  uword *p;
+
+  UPF_TEST (upf_app_add_del (gtm, app_name, 0, 1) == 0, "add app");
+  UPF_TEST (test_add_ip_rule (app_name, 2000,
+                              "permit out ip from 10.10.10.10 to assigned") == 0,
+            "add ip rule 2000");
+  UPF_TEST (test_add_ip_rule (app_name, 2001,
+                              "permit out ip from 192.168.0.0/24 70-100 to assigned") == 0,
+            "add ip rule 2001");
+  UPF_TEST (test_add_ip_rule (app_name, 2002,
+                              "permit out ip from 2001:db8:12::2 to assigned") == 0,
+            "add ip rule 2002");
+  UPF_TEST (test_add_ip_rule (app_name, 2003,
+                              "permit out ip from 2001:db8:13::/64 70-100 to assigned") == 0,
+            "add ip rule 2003");
+  UPF_TEST (upf_app_add_del (gtm, app_name_any, 0, 1) == 0, "add app");
+  UPF_TEST (test_add_ip_rule (app_name_any, 12345,
+                              "permit out ip from any to assigned") == 0,
+            "add ip rule 12345");
+
+  p = hash_get_mem (gtm->upf_app_by_name, app_name);
+  UPF_TEST (!!p, "get app by name");
+  UPF_TEST (!pool_is_free_index (gtm->upf_apps, p[0]), "app entry not free");
+  *app_id = upf_adf_get_adr_db (p[0]);
+  /* UPF_TEST (fib_table_find (FIB_PROTOCOL_IP4, 10000000 + (*app - gtm->upf_apps)) != ~0, */
+  /*           "FIB table created (IP4)"); */
+  /* UPF_TEST (fib_table_find (FIB_PROTOCOL_IP6, 20000000 + (*app - gtm->upf_apps)) != ~0, */
+  /*           "FIB table created (IP6)"); */
+
+  p = hash_get_mem (gtm->upf_app_by_name, app_name_any);
+  UPF_TEST (!!p, "get app by name");
+  UPF_TEST (!pool_is_free_index (gtm->upf_apps, p[0]), "app entry not free");
+  *app_id_any = upf_adf_get_adr_db (p[0]);
+  /* UPF_TEST (fib_table_find (FIB_PROTOCOL_IP4, 10000000 + (*app_any - gtm->upf_apps)) != ~0, */
+  /*           "FIB table created (IP4)"); */
+  /* UPF_TEST (fib_table_find (FIB_PROTOCOL_IP6, 20000000 + (*app_any - gtm->upf_apps)) != ~0, */
+  /*           "FIB table created (IP6)"); */
+
+  vec_free (app_name);
+  vec_free (app_name_any);
+
+  return res;
+}
+
+static int
+cleanup_apps (u32 app_id, u32 app_id_any)
+{
+  int res = 0;
+  upf_main_t * gtm = &upf_main;
+  u8 * app_name = format (0, "IPAPP"), * app_name_any = format (0, "IPANY");
+  /* u32 app_table_id = 10000000 + (app - gtm->upf_apps), */
+  /*   app_any_table_id = 20000000 + (app_any - gtm->upf_apps); */
+
+  upf_adf_put_adr_db (app_id);
+  upf_adf_put_adr_db (app_id_any);
+
+  UPF_TEST (upf_app_add_del (gtm, app_name, 0, 0) == 0, "del IPAPP");
+
+  /* FIXME: trying to remove FIB table causes crash
+  UPF_TEST (fib_table_find (FIB_PROTOCOL_IP4, app_table_id) == ~0,
+            "IPAPP FIB table removed (IP4)");
+  UPF_TEST (fib_table_find (FIB_PROTOCOL_IP6, app_table_id) == ~0,
+            "IPAPP FIB table removed (IP6)");
+  */
+
+  UPF_TEST (upf_app_add_del (gtm, app_name_any, 0, 0) == 0, "del IPANY");
+  /* FIXME: trying to remove FIB table causes crash
```

Comment:
crash caused by assertion of FIB being referenced?
If so, I suppose to skip this case and remove commented code, while leaving just note
If not, and test case assumes FIB to be clear and ready for deletion, probably we'll have to dig into